### PR TITLE
ErrorBoundary subclassing and context

### LINKED ISSUE #33492: ErrorBoundary subclassing, exception context, and recovery
### Description

Cover ...

* DOM events aren't captured by ErrorBoundary.
* The exception context (`@context`).
* Recovery with a clear button (component reference with `Recover`).
* Custom context (`Context={CONTEXT}`).
* Recovery with exception processing (subclassed with overridden `OnErrorAsync`).

`@context` with recovery example ...

```razor
<ErrorBoundary @ref="errorBoundary">
    <ChildContent>
        <EmbeddedCounter />
    </ChildContent>
    <ErrorContent>
        <div class="alert alert-danger" role="alert">
            <p class="fs-3 fw-bold">😈 A rotten gremlin got us. Sorry!</p>
            <p>@context.GetType(): @context.Message</p>
            <button class="btn btn-info" @onclick="_ => errorBoundary.Recover()">Clear</button>
        </div>
    </ErrorContent>
</ErrorBoundary>

@code {
    private ErrorBoundary? errorBoundary;
}
```

Custom context ...

```razor
<ErrorContent Context="exception">
    @exception.Message
</ErrorContent>
```

Subclassed `CustomErrorBoundary` component (`CustomErrorBoundary.razor`) for custom processing with `OnErrorAsync` ...

```razor
@inherits ErrorBoundary
@inject ILogger<CustomErrorBoundary>? Logger

@if (CurrentException is null)
{
    @ChildContent
}
else if (ErrorContent is not null)
{
    @ErrorContent(CurrentException)
}

@code {
    protected override async Task OnErrorAsync(Exception exception)
    {
        await Task.Yield();
        Logger?.LogError("{Type}: {Message}", exception.GetType(), exception.Message);
    }
}
```

***OR*** ... preceding component implemented in code ...

```csharp
using Microsoft.AspNetCore.Components;
using Microsoft.AspNetCore.Components.Web;

namespace BlazorSample;

public class CustomErrorBoundary : ErrorBoundary
{
    [Inject]
    ILogger<CustomErrorBoundary>? Logger {  get; set; }

    protected override async Task OnErrorAsync(Exception exception)
    {
        await Task.Yield();
        Logger?.LogError("{Type}: {Message}", exception.GetType(), exception.Message);
    }
}
```

Used as ...

```razor
<CustomErrorBoundary>
    ...
</CustomErrorBoundary>
```

Cross-references:

* https://github.com/dotnet/aspnetcore/issues/56413#issuecomment-2320385742
* https://github.com/dotnet/aspnetcore/issues/41784
* `<xref:Microsoft.AspNetCore.Components.ErrorBoundaryBase.OnErrorAsync%2A>`

### Page URL

https://learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/handle-errors?view=aspnetcore-8.0

### Content source URL

https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/blazor/fundamentals/handle-errors.md

### Document ID

66a0b1c2-c45c-98ca-9808-6f340a861c44

### Article author

@guardrex

### PR DESCRIPTION
**Pull request by guardrex**
Fixes #33492

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [aspnetcore/blazor/file-uploads.md](https://github.com/dotnet/AspNetCore.Docs/blob/0674980b205e042a5b31b05deca198e3d838b169/aspnetcore/blazor/file-uploads.md) | [ASP.NET Core Blazor file uploads](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/file-uploads?branch=pr-en-us-33585) |
| [aspnetcore/blazor/fundamentals/handle-errors.md](https://github.com/dotnet/AspNetCore.Docs/blob/0674980b205e042a5b31b05deca198e3d838b169/aspnetcore/blazor/fundamentals/handle-errors.md) | [Handle errors in ASP.NET Core Blazor apps](https://review.learn.microsoft.com/en-us/aspnet/core/blazor/fundamentals/handle-errors?branch=pr-en-us-33585) |

### CODE REVIEW

**Review by MackinnonBuck — APPROVED** [review]
Looks good, just a few suggestions 🙂

File: aspnetcore/blazor/fundamentals/handle-errors.md [hunk 1]
```diff
@@ -412,6 +428,86 @@ To avoid the infinite loop where recovering merely rerenders a component that th
...
+
+```razor
+<ErrorBoundary @ref="errorBoundary">
+    <ChildContent>
+        <EmbeddedCounter />
+    </ChildContent>
+    <ErrorContent>
+        <div class="alert alert-danger" role="alert">
+            <p class="fs-3 fw-bold">😈 A rotten gremlin got us. Sorry!</p>
+            <p>@context.GetType(): @context.Message</p>
```
[the remark this answers is not in the document]
Comment (guardrex) [reply aspnetcore/blazor/fundamentals/handle-errors.md:441]: Yes ... I thought about that but didn't state my usual warning about it. I guess I'm looking for ***anything*** that we can do there with the context to demo it. What can it show that would be safe? How about I change it to the `HelpLink`?

File: aspnetcore/blazor/fundamentals/handle-errors.md [hunk 2]
```diff
@@ -412,6 +428,86 @@ To avoid the infinite loop where recovering merely rerenders a component that th
...
+
+public class CustomErrorBoundary : ErrorBoundary
+{
+    [Inject]
+    ILogger<CustomErrorBoundary>? Logger {  get; set; }
+
+    protected override async Task OnErrorAsync(Exception ex)
+    {
+        await Task.Yield();
+        Logger?.LogError("{Type}: {Message}", ex.GetType(), ex.Message);
```
Comment (guardrex) [inline aspnetcore/blazor/fundamentals/handle-errors.md:498]:
```suggestion
        Logger.LogError(ex, "An error occurred while rendering a component.");
        return Task.CompletedTask;
```

File: aspnetcore/blazor/fundamentals/handle-errors.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (guardrex) [reply aspnetcore/blazor/fundamentals/handle-errors.md:441]:
```suggestion
            <p>@context.HelpLink</p>
```

File: aspnetcore/blazor/fundamentals/handle-errors.md [hunk 3]
```diff
@@ -381,12 +381,28 @@ The following example wraps the `EmbeddedCounter` component and supplies custom
 
 For the preceding example, the app's stylesheet presumably includes an `errorUI` CSS class to style the content. The error content is rendered from the <xref:Microsoft.AspNetCore.Components.ErrorBoundaryBase.ErrorContent> property without a block-level element. A block-level element, such as a division (`<div>`) or a paragraph (`<p>`) element, can wrap the error content markup, but it isn't required.
 
+Optionally, use the context (`@context`) of the <xref:Microsoft.AspNetCore.Components.ErrorBoundaryBase.ErrorContent> to obtain error data:
+
+```razor
+<ErrorContent>
+    @context.Message
```
Comment (guardrex) [inline aspnetcore/blazor/fundamentals/handle-errors.md:388]:
```suggestion
    @context.HelpLink
```

File: aspnetcore/blazor/fundamentals/handle-errors.md [hunk 4]
```diff
@@ -381,12 +381,28 @@ The following example wraps the `EmbeddedCounter` component and supplies custom
...
+</ErrorContent>
+```
+
+The <xref:Microsoft.AspNetCore.Components.ErrorBoundaryBase.ErrorContent> can also name the context. In the following example, the context is named `exception`:
+
+```razor
+<ErrorContent Context="exception">
+    @exception.Message
```
Comment (guardrex) [inline aspnetcore/blazor/fundamentals/handle-errors.md:396]:
```suggestion
    @exception.HelpLink
```

File: aspnetcore/blazor/fundamentals/handle-errors.md [hunk 5]
```diff
@@ -412,6 +428,86 @@ To avoid the infinite loop where recovering merely rerenders a component that th
...
+}
+```
+
+You can also subclass <xref:Microsoft.AspNetCore.Components.Web.ErrorBoundary> for custom processing by overriding <xref:Microsoft.AspNetCore.Components.Web.ErrorBoundary.OnErrorAsync%2A>. The following example merely logs the error, but you can implement any error handling code you wish. You can remove the line that awaits <xref:System.Threading.Tasks.Task.Yield%2A?displayProperty=nameWithType> if your code awaits some other asynchronous task.
+
+`CustomErrorBoundary.razor`:
+
+```razor
+@inherits ErrorBoundary
+@inject ILogger<CustomErrorBoundary>? Logger
```
[the remark this answers is not in the document]
Comment (guardrex) [reply aspnetcore/blazor/fundamentals/handle-errors.md]: I checked the whole Blazor node for this mistake ... this is the only instance of me goofing it 🙈👍.

**Pull request merged**